Add checks that optimizers do not modify the layout

julian-elastic opened the pull request:
Add verification that the optimizers do not modify the number of attributes and the attribute datatype.
We add special handling for Lookup Join, by checking EsQueryExec esQueryExec && esQueryExec.indexMode() == LOOKUP and another special handling for ProjectAwayColumns.ALL_FIELDS_PROJECTED

Closes https://github.com/elastic/elasticsearch/issues/125576

julian-elastic committed:
Verify the layout is not changed during optimizations

elasticsearchmachine commented:
Hi @julian-elastic, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

julian-elastic committed:
Update docs/changelog/130855.yaml

elasticsearchmachine commented:
Hi @julian-elastic, I've updated the changelog YAML for you.

julian-elastic committed:
Update docs/changelog/130855.yaml

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Attribute.java:
```diff
@@ -134,4 +134,19 @@ public String nodeString() {
+            return false;
+        }
+        for (int i = 0; i < left.size(); i++) {
+            if (left.get(i).dataType().equals(right.get(i).dataType()) == false) {
```
super nit: since `DataType` is an enum, we typically just use `==` rather than `.equals`.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:
```diff
@@ -13,26 +13,29 @@
 
     /** Verifies the optimized logical plan. */
-    public Failures verify(LogicalPlan plan, boolean skipRemoteEnrichVerification) {
+    public Failures verify(LogicalPlan planAfter, boolean skipRemoteEnrichVerification, LogicalPlan planBefore) {
```
nit/bikeshed: naming could be a little more explicit, maybe:
```suggestion
    public Failures verify(LogicalPlan optimizedPlan, boolean skipRemoteEnrichVerification, LogicalPlan planBeforeOptimization) {
```

Also, since we only need the output of the non-optimized plan, we could consider slimming down the signature to
```
    public Failures verify(LogicalPlan optimizedPlan, boolean skipRemoteEnrichVerification, List<Attribute> expectedOutputAttributes)
```

julian-elastic replied:
Done and done

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:
```diff
@@ -47,6 +50,12 @@ public Failures verify(LogicalPlan plan, boolean skipRemoteEnrichVerification) {
 
+        if (datatypeEquals(planBefore.output(), planAfter.output()) == false) {
+            failures.add(
+                fail(planAfter, "Layout has changed from [{}] to [{}]. ", planBefore.output().toString(), planAfter.output().toString())
```
```suggestion
                fail(planAfter, "Output has changed after optimization from [{}] to [{}]. ", planBefore.output().toString(), planAfter.output().toString())
```
"Layout" is a very specific term that only really exists for physical operators.

julian-elastic replied:
done

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalVerifier.java:
```diff
@@ -67,6 +70,14 @@ public Failures verify(PhysicalPlan plan, boolean skipRemoteEnrichVerification)
         });
 
+        if (datatypeEquals(planBefore.output(), planAfter.output()) == false) {
+            if ((planAfter instanceof EsQueryExec esQueryExec && esQueryExec.indexMode() == LOOKUP) == false) {
```
A small comment maybe, why we can exclude lookups? Something like
```
// LookupJoinExec represents the lookup index with EsSourceExec and this is turned into EsQueryExec by
// ReplaceSourceAttributes. Because InsertFieldExtractions doesn't apply to lookup indices, the
// right hand side will only have the EsQueryExec providing the _doc attribute and nothing else.
```

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalVerifier.java:
```diff
@@ -14,10 +14,13 @@
+import static org.elasticsearch.xpack.esql.core.expression.Attribute.datatypeEquals;
 
 /** Physical plan verifier. */
 public final class PhysicalVerifier {
```
Maybe out of scope: the physical and logical verifiers share quite a bit of code and behavior now; maybe we could have them inherit from the same (abstract) base class by using that `PhysicalPlan` and `LogicalPlan` both inherit from `QueryPlan`. By skimming this, I didn't notice anything that super duper depends on methods of `LogicalPlan` or `PhysicalPlan`.

alex-spies reviewed (commented):
Looks good! Mostly minor remarks.

The only  non-minor remark is on tests: Can we add tests that shows that the verification triggers when optimization changes the plan's output?

One way to test this is using an inherited optimizer instance where we add a final batch which adds a wrong projection on top, which adds/removes some attributes + changes the data types. We could place such a test into the logical and physical optimizer tests, each.

julian-elastic committed:
Address code review feedback part 1

julian-elastic committed:
Address code review feedback part 2

julian-elastic committed:
Merge remote-tracking branch 'origin/outputChecker_v2' into outputChecker_v2

Commit:
[CI] Auto commit changes from spotless

julian-elastic commented:
@alex-spies This is ready for final review, I addressed the code review feedback. I added the UTs to the LocalLogical and LocalPhysical plan optimizer UTs. I am not sure there is any benefit adding them to the Local and Physical plan optimizer UTs as the code is shared.

julian-elastic committed:
Merge branch 'main' into outputChecker_v2

ivancea reviewed (approved):
LGTM! One question about the lookup condition

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PostOptimizationPhasePlanVerifier.java:
```diff
@@ -0,0 +1,71 @@
+            // LookupJoinExec represents the lookup index with EsSourceExec and this is turned into EsQueryExec by
+            // ReplaceSourceAttributes. Because InsertFieldExtractions doesn't apply to lookup indices, the
+            // right hand side will only have the EsQueryExec providing the _doc attribute and nothing else.
+            if ((optimizedPlan instanceof EsQueryExec esQueryExec && esQueryExec.indexMode() == LOOKUP) == false) {
```
We perform an optimizer run on every fragment. LookupJoinExec also contains such a fragment, and currently it only contains an `EsQueryExec` after optimization.

@julian-elastic , I know I'm requesting a lot of comments, but maybe one more wouldn't hurt to explain this?

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:
```diff
@@ -10,30 +10,31 @@
+    @Override
+    void checkPlanConsistency(QueryPlan<?> optimizedPlan, Failures failures, Failures depFailures) {
+        optimizedPlan.forEachUp(p -> {
+            PlanConsistencyChecker.checkPlan(p, depFailures);
```
nit: both the logical and physical verifier have this in `checkPlanConsistency`; couldn't this just go `PostOptimizationPhasePlanVerifier#verify`?

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
Sorry, I missed your comment here:
> I am not sure there is any benefit adding them to the Local and Physical plan optimizer UTs as the code is shared.
I'm not adamant we need them, but I sleep a little better if we have all 4 optimizers tested; while the optimizer code is mostly shared, the code for the verification is in fact not (like [here](https://github.com/elastic/elasticsearch/pull/130855/files#diff-45f172b501f6b73039f2a486c29da73ed84d03468540659d86dfa8e10222ade9L115-R115), the non-local logical optimizer explicitly has to call `verifier.verify(...)`).

julian-elastic committed:
Fix more cases and address code review

alex-spies reviewed (approved):
Looks super good! However, can we merge https://github.com/elastic/elasticsearch/pull/129745 first, so the fix to ProjectAwayColumns is in a separate PR (which we'll backport more aggressively)?

Other than that this LGTM!

julian-elastic committed:
Merge branch 'main' into outputChecker_v2

julian-elastic committed:
Integrate with latest from main and fix merge errors

julian-elastic committed:
Fix a bug due to static counter

julian-elastic committed:
Address more code review comments

julian-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/ProjectAwayColumns.java:
```diff
@@ -82,12 +84,13 @@ public PhysicalPlan apply(PhysicalPlan plan) {
                             output = Expressions.asAttributes(fields);
                         }
                         // add a logical projection (let the local replanning remove it if needed)
+                        output = reorderOutput(logicalFragment.output(), output);
```
Done, integrated with https://github.com/elastic/elasticsearch/pull/129745

julian-elastic committed:
Fix merge error

alex-spies reviewed (approved):
Looking good, let's :shipit: !
